kernel: validate kernel objects

tyrantlizard77 opened the pull request:
This will be used in conjunction with the kernel/userspace
privilege split, when a user thread makes a system call with
a kernel object as a parameter, the validity of the kernel
object pointer passed in will be confirmed, and cause a kernel
oops if it wasn't valid.

This is done by randomly generating a set of XOR keys for each
kernel object type, and when an object is initialized its
pointer value will be XOR'd with the key and stored within
the object itself. When an object is being validated,
XORing this value again with the key for the object type should
pop out the original pointer again.

We validate most, but not all kernel objects. We currently
omit k_mbox, k_queue, k_mem_pool as they appear to need some
redesign before we can use them from userspace due to how they
store data. JIRAs have been opened for each of these.

Issue: ZEP-2187
Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
kernel: validate kernel objects

This will be used in conjunction with the kernel/userspace
privilege split, when a user thread makes a system call with
a kernel object as a parameter, the validity of the kernel
object pointer passed in will be confirmed, and cause a kernel
oops if it wasn't valid.

This is done by randomly generating a set of XOR keys for each
kernel object type, and when an object is initialized its
pointer value will be XOR'd with the key and stored within
the object itself. When an object is being validated,
XORing this value again with the key for the object type should
pop out the original pointer again.

We validate most, but not all kernel objects. We currently
omit k_mbox, k_queue, k_mem_pool as they appear to need some
redesign before we can use them from userspace due to how they
store data. JIRAs have been opened for each of these.

Issue: ZEP-2187
Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
I want to to turn this on by default for all the QEMU targets, but waiting on some fixes for ZEP-2397 before I can do that.

lpereira reviewed (commented):
I still have to review this thoroughly, but there are some nits.

andrewboie commented:
rebased and addressed @lpereira review comments

Commit:
boards: enable object verification for QEMU builds

This will help catch bugs when running sanitycheck if an
application or subsystem tries to make kernel calls on
something that is not a valid kernel object.

This adds a bit of overhead but we don't gather performance
metrics in QEMU anyway, it's too dependent on the
characteristics of the host system.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andyross commented on kernel/object_validate.c:
```diff
@@ -0,0 +1,224 @@
+
+	initialized = 1;
+
+	for (i = 0; i < K_OBJ_LAST; i++) {
```
Don't see why you can't fold these into the same loop.  Initialize the metadata, then loop over the defined objects.

andrewboie commented:
Addressed @andyross feedback and added a more helpful error message if an app defines a kernel object in non-kernel memory.

andyross commented on kernel/object_validate.c:
```diff
@@ -0,0 +1,238 @@
+		struct object_metadata *o = &ometa[i];
+
+#if CONFIG_RANDOM_HAS_DRIVER
+		o->xor_key = sys_rand32_get();
```
Thinking about this, we don't want a random number here.  Think of the requirements:

1. It should be "white" relative to the the space of pointers and the other xor_key values.  A change in any bit of the pointer should be ~50% likely to invalidate the match.  It shouldn't be "too close" to the other keys.

2. It should not be the same as any other xor_key

A RNG will meet those only statistically.  And worse, because it's runtime-dependent (on systems with hardware entropy anyway) we're pretty much guaranteed that eventually a system will boot up having hit the birthday jackpot and have colliding (or even partially colliding, which is still bad) keys.

We can do better by precomputing it at build time.  A quick python script that does the RNG and tests for validity before emitting C code to define the ometa[] array would be easy and simplify the boot code (heh) to boot.

andrewboie replied:
@andyross good point, but I fear static keys only makes the situation worse, as soon as someone discovers the set of static XOR keys by just reading them out the flash, or JTAG, or whatever, that will enable every script kiddie to hack every single device out there.

not seeing a good way out of this :/

andyross replied:
Well, yeah.  It's not really a security feature anyway.  But the complaint above was about robustness, it's possible to boot in a configuration where the validity check doesn't work right.

FWIW: you could do both: keep a single/global per-boot randomized key and a per-type value and xor with both (potentially pre-computing the xor_key values in place, assuming they're in RAM as they are now).  Probably not worth it.

andrewboie replied:
If we admit that this is totally insecure we might as well not have random numbers at all, and just use the loop counter (basically just the !CONFIG_RANDOM_HAS_DRIVER case) maybe with some math done on it to ensure not too close to other keys.

i guess let this patch be a prototype of the pointer-based API model. i'm going to prototype in another patch what a descriptor-based validation mechanism would look like, I think it will at least inform the discussion.

inakypg replied:
You could do multiple layers very easily
You could have keys per thread, per boot and per type ... and they all have to match in the right order. That significantly increases the difficulty of making a guess.

andrewboie replied:
@inakypg that falls apart if you need to grant more than one thread access. if you grant exactly one thread access, sure you can just xor its struct k_thread pointer in. but multiple threads, not gonna work, you can't just XOR in all their pointers in. in practice we are going to need to grant multiple threads access to some object all the time.

inakypg replied:
For the record so we don't drive anyone crazy:

https://jira.zephyrproject.org/browse/ZEP-2408?focusedCommentId=19363&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-19363

where we are having that discussion :)

andrewboie commented:
Investigating a different approach. Closing for now.
